flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ZooKeeperCompletedCheckpointStore.java

```diff
@@ -226,16 +200,43 @@ public CompletedCheckpoint getLatestCheckpoint() throws Exception {
 			return null;
 		}
 		else {
-			return checkpointStateHandles.getLast().f0.retrieveState();
+			while(!checkpointStateHandles.isEmpty()) {
+				Tuple2<RetrievableStateHandle<CompletedCheckpoint>, String> checkpointStateHandle = checkpointStateHandles.peekLast();
+
+				try {
+					return retrieveCompletedCheckpoint(checkpointStateHandle);
+				} catch (FlinkException e) {
```

Comment:
Technically, I think it was ok, because the `retrieveCompletedCheckpoint` method catches all `Exceptions` and wraps them in a `FlinkException`. But it's better to not rely on this implementation detail.